### PR TITLE
Refactor Retry test util to not use CompletableFuture

### PR DESCRIPTION
**Pull request by thecoop**
This simplifies `Retry` test utility class to not use `CompletableFuture`, and also ensures the created thread stops in all situations

### CODE REVIEW

**Comment by williamrandolph** [comment]
@thecoop Why was `CompletableFuture` an inappropriate class for what we want to do here? At first glance it seems like it fits the use case. Apologies if I'm just missing the context of other PRs.

**Comment by thecoop** [comment]
This is mainly just to simplify it, this is using `CompletableFuture` to start a thread up in a roundabout way, where we could just create the thread directly. This change also makes sure the thread exits in all cases, the previous version didn't do this.

**Review by williamrandolph — COMMENTED** [review]
I definitely support making sure that the thread exits in all cases, but I wonder if some other dev might come along, see that we're directly managing a thread, and think that it might be simpler to use Java's higher-level async execution tools.

I guess the argument would be: `CompletableFuture` offers the ability to combine and chain asynchronous executions, but since all we want to do is retry an operation on a single thread and handle exceptions, using `CompletableFuture` and making a custom executor is unnecessary overhead.

Basically, my nitpick is that I think it'd be helpful to have a brief code comment so devs can understand why we're managing a thread directly via `start`/`join` rather than using higher-level Java classes.

**Comment by thecoop** [comment]
I was trying to keep the same external semantics - but this isn't a very used class. Callers also do some filesystem & process stuff that might be affected by different threads...?

**Review by williamrandolph — APPROVED** [review]
At this point I think I'm good to give my checkmark. I doubt that anyone will restore the `CompletableFuture` for no reason, and if there is a reason we need to use that class, we can talk about it in that future PR. Or, as Ryan said, we may be cleaning up this class soon, making the whole point moot.

LGTM, especially insofar as it makes sure we always exit the thread.

**Comment by thecoop** [comment]
This was just a random change I spotted doing a bit of code gardening, if its not worth bothering about I'll just close this
